### ISSUE TITLE
Don't build .app bundle when "buildable:False"

### ISSUE THREAD

**Issue by HeinrichApfelmus** [gimbo/cabal-macosx#7]
Hello!

I have a .cabal file with the following example executable.

```
Executable Asteroids
    if flag(buildExamples)
        build-depends:
            random == 1.0.*,
            executable-path == 0.0.*,
            filepath >= 1.1 && <= 1.4
    else
        buildable: False
    hs-source-dirs:  src
    other-modules:   Paths_reactive_banana_wx
    main-is:         Asteroids.hs
```

When the `buildable` flag is false, the executable won't be built, but cabal-macosx still makes an (empty) .app bundle. This is undesirable and takes up a lot of time. Any fix?

**Comment by gimbo** [comment]
Thanks for the report, Heinrich - I'll look into it now.

**Comment by gimbo** [comment]
I've just pushed a commit which might fix this - please could you test, Heinrich?

I say "might", because my (admittedly cursory) attempts to reproduce your problem didn't succeed.  I tried modifying examples/wxHello/wxHello.cabal along the lines showed above, but when the relevant flag isn't set, rather than seeing the behaviour you describe, I just see:

```
setup: Package wxHello-0.1.0 can't be built on this system.
```

So please could you test, and if it's not fixed, please could you show me a modified wxHello.cabal which exhibits the problem bahviour?

Thanks!

**Comment by HeinrichApfelmus** [comment]
I am so very stupid, I should have mentioned that I wrote a custom `Setup.hs` script! Sorry!

Ok, so I have checked out your latest commit, cabal installed it and built the package again, which has the following [.cabal file](https://github.com/HeinrichApfelmus/reactive-banana/blob/master/reactive-banana-wx/reactive-banana-wx.cabal) and [custom Setup.hs](https://github.com/HeinrichApfelmus/reactive-banana/blob/master/reactive-banana-wx/Setup.hs).

After cleaning out my `dist` directory, I get

```
$ rm -R dist

$ cabal configure
Resolving dependencies...
[1 of 1] Compiling Main             ( Setup.hs, dist/setup/Main.o )
Linking ./dist/setup/setup ...
Configuring reactive-banana-wx-0.6.0.0...

$ cabal build
Preprocessing library reactive-banana-wx-0.6.0.0...
Preprocessing executables for reactive-banana-wx-0.6.0.0...
Building reactive-banana-wx-0.6.0.0...
[1 of 1] Compiling Reactive.Banana.WX ( src/Reactive/Banana/WX.hs, dist/build/Reactive/Banana/WX.o )
[1 of 1] Compiling Reactive.Banana.WX ( src/Reactive/Banana/WX.hs, dist/build/Reactive/Banana/WX.p_o )
Registering reactive-banana-wx-0.6.0.0...
Creating application bundle directory dist/build/Asteroids.app
Copying executable Asteroids into place
setup: dist/build/Asteroids/Asteroids: does not exist
```

So, basically, the problem is that the `appBundleBuildHook` is executed unconditionally in the `postBuild` hook, even when no executable was actually built. I'm don't know enough about cabal to fix this, but the solution is probably to have `appBundleBuildHook` package executables on when they were built? Or maybe I have to change my custom Setup.hs script, but I don't know how.

**Comment by HeinrichApfelmus** [comment]
Ok, I can reproduce the behavior with `wxHello`. The trick is to put the `cabal-macosx` dependency in a library. Here the .cabal file.

```
Name:                   wxHello
Version:                0.1.0
Stability:              Alpha
Synopsis:               wxWidgets `Hello World' example for cabal-macosx
Description:
    Example showing how to use cabal-macosx to build an application
    bundle for a simple `Hello World' program using the wxWidgets GUI
    toolkit.
Category:               Data
License:                BSD3
License-file:           LICENSE
Copyright:              Andy Gimblett <haskell@gimbo.org.uk>
Author:                 Andy Gimblett <haskell@gimbo.org.uk>
Maintainer:             Andy Gimblett <haskell@gimbo.org.uk>
Build-Type:             Custom
Cabal-Version:          >=1.6

Library
    hs-source-dirs:  src
    build-depends:   base >= 3 && < 5, cabal-macosx, wx
    exposed-modules: Main

flag buildExamples
  description: Build example executables
  default:     False

Executable WxHello
    if flag(buildExamples)
        buildable: True
    else
        buildable: False
  hs-source-dirs:       src
  Main-is:              Main.hs
  ghc-options:          -fwarn-tabs -threaded -Wall
```

**Comment by gimbo** [comment]
Thanks Heinrich - that's really helpful! I think I've nailed it now.  I came close before, but not quite.  If no MacApps are explicitly listed in Setup.hs, cabal-macosx tries to build an app for all executables from the .cabal file, and my previous "fix" filtered that list on buildability - but I'd failed to filter in the case where MacApps _are_ explicitly listed in Setup.hs; have now fixed that and the wxHello example seems to work sensibly.  Please could you check your larger example, and if we're good, I'll make a new point release. Thanks again!

**Comment by HeinrichApfelmus** [comment]
After doing `cabal clean` to get rid of the old `Setup` executable, it works like a charm now.

```
No buildable MacApps, so making no bundles.
```

Thanks a lot!

(Feel free to close the issue.)

**Comment by gimbo** [comment]
Great - thanks very much for pointing this out!

-Andy

**Comment by HeinrichApfelmus** [comment]
Dear Andy,

could you make a minor release on hackage with the fixes that you've already made? I just realized that reactive-banana-wx-0.7 currently doesn't always install on Mac OS X due to this issue.
